hybridse/src/udf/default_defs/window_functions_def.cc

```diff
@@ -90,6 +92,105 @@ node::ExprNode* BuildAt(UdfResolveContext* ctx, ExprNode* input, ExprNode* idx,
     return res;
 }
 
+template <typename T>
+struct NthValueWhere {
+    // C type for value
+    using CType = typename DataTypeTrait<T>::CCallArgType;
+
+    void operator()(UdafRegistryHelper& helper) {  // NOLINT
+        helper.library()
+            ->RegisterUdafTemplate<Impl>(helper.name())
+            .doc(helper.GetDoc())
+            .template args_in<int16_t, int32_t, int64_t>();
+    }
+
+    template <typename NthType>
+    struct Impl {
+        void operator()(UdafRegistryHelper& helper) {  // NOLINT
+            std::string prefix = absl::StrCat(helper.name(), "_", DataTypeTrait<T>::to_string(), "_",
+                                              DataTypeTrait<NthType>::to_string());
+            helper.templates<Nullable<T>, Opaque<ContainerT>, Nullable<T>, NthType, Nullable<bool>>()
+                .init(prefix + "_init", Init)
+                .update(prefix + "_update", Update)
+                .output(prefix + "_output", reinterpret_cast<void*>(Output), true);
+        }
+
+     private:
+        struct ContainerT {
+            ContainerT() : nth(0), cur_pos(0) {}
+
+            // given nth value, be negative or positive
+            NthType nth;
+
+            // current iterator pos (start from 1), non-negative
+            NthType cur_pos;
+
+            // saved value list, only updated if `nth` > 0
+            std::vector<std::pair<T, bool>> data;
+        };
+
+        // (nth idx, [(value, value_is_null)])
+        // using ContainerT = std::pair<NthType, std::vector<std::pair<T, bool>>>;
+
+        static void Init(ContainerT* ctr) { new (ctr) ContainerT(); }
+
+        static ContainerT* Update(ContainerT* ctr, CType value, bool value_is_null, NthType nth, bool cond,
+                                  bool cond_is_null) {
+            // by default, update is iterated from window end to window start
+            if (ctr->nth == 0) {
+                // update only once, we treat nth as constant
+                ctr->nth = nth;
+            }
+            if (cond && !cond_is_null) {
+                if (ctr->nth > 0) {
+                    // nth from window start
+                    ctr->data.emplace_back(container::ContainerStorageTypeTrait<T>::to_stored_value(value),
```

Comment:
a fixed-size buffer like dequeue or list (push_back and pop_front) may work well

Reply:
sounds good, I impl a simple ring buffer for this